forward/fwd.go

```diff
@@ -206,22 +206,39 @@ func (f *httpForwarder) serveHTTP(w http.ResponseWriter, req *http.Request, ctx
 	}
 }
 
+func (f *httpForwarder) getURLFromRequest(req *http.Request, ctx *handlerContext) *url.URL {
+	// If the Request was created by Go via a real HTTP request,  RequestURI will
+	// contain the original query string. If the Request was created in code, RequestURI
+	// will be empty, and we will use the URL object instead
+	u := req.URL
+	if req.RequestURI != "" {
+		parsedURL, err := url.ParseRequestURI(req.RequestURI)
+		if err == nil {
+			u = parsedURL
+		} else {
+			ctx.log.Warningf("gravitational/oxy/forward: error when parsing RequestURI: %s", err)
+		}
+	}
+	return u
+}
+
 // copyRequest makes a copy of the specified request to be sent using the configured
 // transport
-func (f *httpForwarder) copyRequest(req *http.Request, u *url.URL) *http.Request {
+func (f *httpForwarder) copyRequest(req *http.Request, target *url.URL, ctx *handlerContext) *http.Request {
 	outReq := new(http.Request)
 	*outReq = *req // includes shallow copies of maps, but we handle this below
 
-	outReq.URL = utils.CopyURL(req.URL)
-	outReq.URL.Scheme = u.Scheme
-	outReq.URL.Host = u.Host
-	outReq.URL.Opaque = req.RequestURI
-	// raw query is already included in RequestURI, so ignore it to avoid dupes
-	outReq.URL.RawQuery = ""
-	// Do not pass client Host header unless optsetter PassHostHeader is set.
-	if !f.passHost {
-		outReq.Host = u.Host
-	}
+	outReq.URL = utils.CopyURL(outReq.URL)
+	outReq.URL.Scheme = target.Scheme
+	outReq.URL.Host = target.Host
+
+	u := f.getURLFromRequest(outReq, ctx)
+
+	outReq.URL.Path = u.Path
+	outReq.URL.RawPath = u.RawPath
+	outReq.URL.RawQuery = u.RawQuery
+	outReq.RequestURI = "" // Outgoing request should not have RequestURI
```

Comment:
So as far as I understand, the actual fix is instead of setting the entire request URI as "opaque" to parse the URL and set parts like path/query individually, correct?

If so, what does the `getURLFromRequest` change fix? Which scenario we could be using incorrect URL in? Just want to make sure we understand why this change is needed so we don't break anything.

Reply:
FWIW none of this is actually my code save for massaging it into our fork -- this is all from upstream. I'm unsure what the `getURLFromRequest` call was originally introduced for, but it sounds like it's a way of ensuring the URL is retrieved properly from the request regardless of how it's created, either through programmatic creation of an `http.Request` or sending the request through an actual http client.

Reply:
https://github.com/vulcand/oxy/pull/108 Is the relevant PR with sadly very little description.

Reply:
@r0mant And I just realized I didn't answer your first question. Yes, that's correct, not using `Opaque` and doing an explicit parse fixes this.

Reply:
Looking at the linked PR and the comment, seems like this case was specifically for the test which creates `http.NewRequest` itself so we shouldn't really hit it in production.